update sync_diff_inspector: sync_diff_inspector v2.0 (#6774)
pingcap/docs: sync-diff-inspector/sync-diff-inspector-overview.md

* Support [data check for TiDB upstream-downstream clusters](/sync-diff-inspector/upstream-downstream-diff.md)
* Support [data check in the DM replication scenario](/sync-diff-inspector/dm-diff.md)

## Restrictions of sync-diff-inspector

* Online check is not supported for data migration between MySQL and TiDB. Ensure that no data is written into the upstream-downstream checklist, and that data in a certain range is not changed. You can check data in this range by setting `range`.

* `JSON`, `BIT`, `BINARY`, `BLOB` and other types of data are not supported. When you perform a data check, you need to set `ignore-columns` to skip checking these types of data.

* In TiDB and MySQL, `FLOAT`, `DOUBLE` and other floating-point types are implemented differently. `FLOAT` and `DOUBLE` respectively take 6 and 15 significant digits for calculating checksum. If you do not want to use this feature, set `ignore-columns` to skip checking these columns.

* Support checking tables that do not contain the primary key or the unique index. However, if data is inconsistent, the generated SQL statements might not be able to repair the data correctly.

## Database privileges for sync-diff-inspector

> sync-diff-inspector needs to obtain the information of table schema to query data. The required database privileges are as follows:

Author: ```suggestion
sync-diff-inspector needs to obtain the information of table schema and to query data. The required database privileges are as follows:
```

* Upstream database
    - `SELECT` (checks data for comparison)
    - `SHOW_DATABASES` (views database name)
    - `RELOAD` (views table schema)
* Downstream database
    - `SELECT` (checks data for comparison)
    - `SHOW_DATABASES` (views database name)
    - `RELOAD` (views table schema)

## Configuration file description

The configuration of sync-diff-inspector consists of the following parts:

- `Global config`: General configurations, including number of threads to check, whether to export SQL statement to fix inconsistent tables, whether to campare the data and so on.
- `Databases config`: Configures the instances of the upstream and downstream databases.
- `Tables config`: Special configurations for specific tables, including specified ranges, columns to be ignored and so on (optional).

> - `Routes`: Rules for upstream multiple schema names to regularly match downstream single schema names (optional).

Author: ```suggestion
- `Routes`: Rules for upstream multiple schema names to match downstream single schema names (optional).
```
```suggestion
- `Routes`: Rules for upstream multiple schema names to regularly match downstream single schema names (optional).
```

- `Task config`: Configures the tables for checking. If some tables have a certain mapping relationship between the upstream and downstream databases or have some special requirements, you must configure these tables.

Below is the description of a complete configuration file:

``` toml
# Diff Configuration.

######################### Global config #########################

> # The number of goroutines created to check data. The number of connections between upstream and downstream databases are slightly greater than this value.

Author: ```suggestion
# The number of goroutines created to check data. The number of connections between sync-diff-inspector and upstream/downstream databases is slightly greater than this value.
```

check-thread-count = 4

# If enabled, SQL statements is exported to fix inconsistent tables.
export-fix-sql = true

# Only compares the table structure instead of the data.
check-struct-only = false

######################### Datasource config #########################
[data-sources]

> [data-sources.mysql1] # mysql1 is the only ID for the database instance. It is used in the following `task.source-instances/task.target-instance` files.

Author: ```suggestion
[data-sources.mysql1] # mysql1 is the only ID for the database instance. It is used for following `task.source-instances/task.target-instance` configuration.
```

    host = "127.0.0.1"
    port = 3306
    user = "root"
    password = ""

[data-sources.tidb0]
    host = "127.0.0.1"
    port = 4000
    user = "root"
    password = ""
    # Uses the snapshot feature. If enabled, historical data is used for comparison
    # snapshot = "386902609362944000"

######################### Table config #########################
# Special configurations for specific tables. The tables to be configured must be in `task.target-check-tables`.

> [table-configs.config1] # config1  is the only ID for this configuration. It is used in the following `task.target-configs`.

Author: ```suggestion
[table-configs.config1] # config1  is the only ID for this configuration. It is used for the following `task.target-configs` configuration.
```

[task]
    # output-dir saves the following information:
    # 1 sql: The SQL file to fix tables that is generated after error is detected. One chunk corresponds to one SQL file.
    # 2 log: sync-diff.log
    # 3 summary: summary.txt
    # 4 checkpoint: a dir
    output-dir = "./output"
    # The upstream database. The value is the unique ID declared by data-sources.
    source-instances = ["mysql1"]
    # The downstream database. The value is the unique ID declared by data-sources.
    target-instance = "tidb0"
    # The tables of downstream databases to be compared. Each table needs to contain schema name and table name, separated by '.'
    # Use "?" to match any character and “*” to match characters of any length.
    # For detailed match rules, refer to golang regexp pkg: https://github.com/google/re2/wiki/Syntax.
    target-check-tables = ["schema*.table*", "!c.*", "test2.t2"]
    # Extra configurations for some tables.
    target-configs= ["config1"]
```

## Run sync-diff-inspector

Run the following command:

{{< copyable "shell-regular" >}}

```bash
./bin/sync_diff_inspector --config=./config.toml
```

> This command outputs a check report `summary.txt` in the `output-dir` of `config.toml` and the log `sync_diff.log`. In the `output-dir`, a folder named by the hash value of the `config. toml` file are also generated. This folder includes the checkpoint node information of breakpoint and the SQL file generated when the data is inconsistent.

Reviewer: ```suggestion
This command outputs a check report `summary.txt` in the `output-dir` of `config.toml` and the log `sync_diff.log`. In the `output-dir`, a folder named by the hash value of the `config. toml` file is also generated. This folder includes the checkpoint node information of breakpoints and the SQL file generated when the data is inconsistent.
```

Comparing the table structure of ``sbtest`.`sbtest96`` ... equivalent
Comparing the table structure of ``sbtest`.`sbtest99`` ... equivalent
Comparing the table data of ``sbtest`.`sbtest96`` ... failure
Comparing the table data of ``sbtest`.`sbtest99`` ...
_____________________________________________________________________________
Progress [==========================================================>--] 98% 193/200
```

```progress
A total of 2 tables need to be compared

Comparing the table structure of ``sbtest`.`sbtest96`` ... equivalent
Comparing the table structure of ``sbtest`.`sbtest99`` ... equivalent
Comparing the table data of ``sbtest`.`sbtest96`` ... failure
Comparing the table data of ``sbtest`.`sbtest99`` ... failure
_____________________________________________________________________________
Progress [============================================================>] 100% 0/0
The data of `sbtest`.`sbtest99` is not equal
The data of `sbtest`.`sbtest96` is not equal

The rest of tables are all equal.
The patch file has been generated in
        'output/fix-on-tidb2/'
You can view the comparision details through 'output/sync_diff.log'
```

### Output file

> The output file directory structure is as follows:

Reviewer: ```suggestion
The directory structure of the output file is as follows:
```

```
output/
|-- checkpoint # Saves the breakpoint information
| |-- bbfec8cc8d1f58a5800e63aa73e5 # Config hash. The placeholder file which identifies the configuration file corresponding to the output directory (output/)
│ |-- DO_NOT_EDIT_THIS_DIR
│ └-- sync_diff_checkpoints.pb # The breakpoint information
|
|-- fix-on-target # Saves SQL files to fix data inconsistency
| |-- xxx.sql
| |-- xxx.sql
| └-- xxx.sql
|
|-- summary.txt # Saves the summary of the check results
└-- sync_diff.log # Saves the output log informatiion when sync-diff-inspector is running
```

### Log

The log of sync-diff-inspector is saved in `${output}/sync_diff.log`, among which `${output}` is the value of `output-dir` in the `config.toml` file.

### Progress

The running sync-diff-inspector periodically (every 10 seconds) prints the progress in checkpoint, which is located at `${output}/checkpoint/sync_diff_checkpoints.pb`, among which `${output}` is the value of `output-dir` in the `config.toml` file.

### Result

> After the check is finished, sync-diff-inspector outputs a report, which is located at `${output}/summary.txt`, among which `${output}` is the value of `output-dir` in the `config.toml` file.

Reviewer: ```suggestion
After the check is finished, sync-diff-inspector outputs a report. It is located at `${output}/summary.txt`, and `${output}` is the value of `output-dir` in the `config.toml` file.
```

```summary
+---------------------+--------------------+----------------+
|        TABLE        | STRUCTURE EQUALITY | DATA DIFF ROWS |
+---------------------+--------------------+----------------+
| `sbtest`.`sbtest99` | true               | +97/-97        |
| `sbtest`.`sbtest96` | true               | +0/-101        |
+---------------------+--------------------+----------------+
Time Cost: 16.75370462s
Average Speed: 113.277149MB/s
```

- TABLE: The corresponding database and table names

- STRUCTURE EQUALITY: Checks whether the table structure is the same

- DATA DIFF ROWS: `rowAdd` / `rowDelete`. Indicates the number of rows that need to be added/deleted to fix the table

### SQL statements to fix inconsistent data

> If different rows exist during the data checking process, the SQL statements to fix them will generate. If the data inconsistency exists in a chunk, a SQL file named by `chunk.Index` will generate. The SQL file is located at `${output}/fix-on-${instance}`, among which `${instance}` is the value of `task.target-instance` in the `config.toml` file.

Reviewer: ```suggestion
If different rows exist during the data checking process, the SQL statements will be generated to fix them. If the data inconsistency exists in a chunk, a SQL file named by `chunk.Index` will be generated. The SQL file is located at `${output}/fix-on-${instance}`, and `${instance}` is the value of `task.target-instance` in the `config.toml` file.
```

A SQL file contains the tale to which the chunk belong and the range information. For the SQL files, you should consider the following three situations:

- If the rows in the downstream database are missing, REPLACE statements will be applied
- If the rows in the downstream database are redundant, DELETE statements will be applied

> - If some data of the rows in the downstream database is inconsistent, REPLACE statements will be applied and different columns will be marked with annotation in the SQL file

Reviewer: ```suggestion
- If some data of the rows in the downstream database is inconsistent, REPLACE statements will be applied and inconsistent columns will be marked with annotation in the SQL file
```

  target data  ╏ 5003616 ╏ '0709824117-9809973320-4456050422' ╏ '1714066100-7057807621-1425865505'
╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╋╍╍╍╍╍╍╍╍╍╋╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╋╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍╍
*/
REPLACE INTO `sbtest`.`sbtest99`(`id`,`k`,`c`,`pad`) VALUES (3700000,2501808,'hello','world');
```

## Note

- sync-diff-inspector consumes a certain amount of server resources when checking data. Avoid using sync-diff-inspector to check data during peak business hours.
- TiDB uses the `utf8_bin` collation. If you need to compare the data in MySQL with that in TiDB, pay attention to the collation configuration of MySQL tables. If the primary key or unique key is the `varchar` type and the collation configuration in MySQL differs from that in TiDB, then the final check result might be incorrect because of the collation issue. You need to add collation to the sync-diff-inspector configuration file.
- sync-diff-inspector divides data into chunks first according to TiDB statistics and you need to guarantee the accuracy of the statistics. You can manually run the `analyze table {table_name}` command when the TiDB server's *workload is light*.

> - Pay special attention to `table-rules`. If you configure `schema-pattern="test1"`, `table-pattern = "t_1"`, `target-schema="test2"` and `target-table = "t_2"`, the `test1`.`t_1` schema in the source database and the `test2`.`t_2` schema in the target database are compared. Sharding is enabled by default in sync-diff-inspector, so if the source database has a `test2`.`t_2` table,  the `test1`.`t_1` table and `test2`.`t_2` table in the source database serving as sharding are compared with the `test2`.`t_2` table in the target database.

Reviewer: 出现这种问题会对用户有什么影响呢？如果有风险，可以加个说明明确指出。

Author: @Leavrth PTAL~

Reviewer 2: 一般来说，如果没有rules，那么上游数据库会匹配下游数据库相同名称的表，相当于每个表都有个默认的自己匹配自己的rule。但是加上rules并不会去掉这个默认的匹配规则，也就是如上面说的，如果用户想要做 test1.t_1 -> test2.t_2 的匹配，但是默认还存在 test2.t_2 -> test2.t_2 的匹配，那么上游存不存在 test2.t_2 的结果是不一样的。
